### PR TITLE
[IMP] delivery: carrier is not set after delivery generated

### PR DESCRIPTION
**Pull request by kir-odoo**
Before this commit
==================
when SO is confirmed and Delivery is created then add the shipping method
in SO, in this case, the shipping carrier is not set in the existing undelivered
delivery of that SO.

After this commit
=================
So in this commit, we set the shipping carrier for that delivery.

TaskId: 2946360

### CODE REVIEW

File: addons/delivery/models/sale_order.py [hunk 1]
```diff
@@ -51,9 +51,13 @@ def _remove_delivery_line(self):
         to_delete.unlink()
 
     def set_delivery_line(self, carrier, amount):
+        # Prevent copy of the carrier and carrier price when generating return picking
```
Comment (jva-odoo) [inline addons/delivery/models/sale_order.py:54]: @kir-odoo update comment msg doesn't just copy-paste.

File: addons/delivery/tests/test_delivery_stock_move.py [hunk 2]
```diff
@@ -184,3 +184,57 @@ def test_03_invoiced_status(self):
...
+        delivery_wizard = Form(self.env['choose.delivery.carrier'].with_context({
+            'default_order_id': self.sale_prepaid.id,
+            'default_carrier_id': self.normal_delivery.id,
+        }))
+        choose_delivery_carrier = delivery_wizard.save()
+        choose_delivery_carrier.button_confirm()
+
+        # Check the carrier in picking after conform sale order
+        self.assertEqual(self.sale_prepaid.picking_ids.carrier_id, self.normal_delivery, 'The shipping method should be specified in pickings too.')
+        self.assertEqual(return_picking.carrier_id.id, False, 'The shipping method should not set in return')
```
Comment (jva-odoo) [inline addons/delivery/tests/test_delivery_stock_move.py:240]:
```suggestion
        self.assertEqual(return_picking.carrier_id.id, False, 'The shipping method should not set in return picking')
```

File: addons/delivery/tests/test_delivery_stock_move.py [hunk 3]
```diff
@@ -184,3 +184,57 @@ def test_03_invoiced_status(self):
...
+        wiz_act = self.sale_prepaid.picking_ids.button_validate()
+        Form(self.env[wiz_act['res_model']].with_context(wiz_act['context'])).save().process()
+
+        # Return picking
+        return_form = Form(self.env['stock.return.picking'].with_context(active_id=self.sale_prepaid.picking_ids.id, active_model='stock.picking'))
+        return_wizard = return_form.save()
+        action = return_wizard.create_returns()
+        return_picking = self.env['stock.picking'].browse(action['res_id'])
+
+        #add new product
```
Comment (jva-odoo) [inline addons/delivery/tests/test_delivery_stock_move.py:219]:
```suggestion
        #add new product so new picking is created
```

File: addons/delivery/tests/test_delivery_stock_move.py [hunk 4]
```diff
@@ -184,3 +184,57 @@ def test_03_invoiced_status(self):
...
+        self.sale_prepaid.write({
+            'order_line': [(0, 0, {
+                'name': 'Cable Management Box',
+                'product_id': self.product_cable_management_box.id,
```
Comment (jva-odoo) [inline addons/delivery/tests/test_delivery_stock_move.py:223]:
```suggestion
                'name': 'Another product to deliver',
                'product_id': self.product_11.id,
```

File: addons/delivery/tests/test_delivery_stock_move.py [hunk 5]
```diff
@@ -184,3 +184,57 @@ def test_03_invoiced_status(self):
         choose_delivery_carrier.button_confirm()
 
         self.assertEqual(so.invoice_status, 'no', 'The status should still be "Nothing To Invoice"')
+
+    def test_delivery_carrier_from_confirm_so(self):
+        '''Test if add shipping method in sale order after conform
+           will it add in pickings too or not"'''
+
+        self.sale_prepaid = self.SaleOrder.create({
```
Comment (jva-odoo) [inline addons/delivery/tests/test_delivery_stock_move.py:192]:
```suggestion
        sale_prepaid = self.SaleOrder.create({
```

File: addons/delivery/tests/test_delivery_stock_move.py [hunk 6]
```diff
@@ -184,3 +184,57 @@ def test_03_invoiced_status(self):
...
+        # Add delivery cost in Sales order
+        delivery_wizard = Form(self.env['choose.delivery.carrier'].with_context({
+            'default_order_id': self.sale_prepaid.id,
+            'default_carrier_id': self.normal_delivery.id,
+        }))
+        choose_delivery_carrier = delivery_wizard.save()
+        choose_delivery_carrier.button_confirm()
+
+        # Check the carrier in picking after conform sale order
+        self.assertEqual(self.sale_prepaid.picking_ids.carrier_id, self.normal_delivery, 'The shipping method should be specified in pickings too.')
```
Comment (jva-odoo) [inline addons/delivery/tests/test_delivery_stock_move.py:239]:
```suggestion
       delivery_of_product_11 = sale_prepaid.picking_ids.filtered(lambda p: self.product_11 in p.move_ids.product_id)
        self.assertEqual(delivery_for_product_11.carrier_id, self.normal_delivery, 'The shipping method should be set in pending deliveries.')
        done_delivery = sale_prepaid.picking_ids.filtered(lambda p: p.state == "done")
        self.assertEqual(done_delivery.carrier_id, False, 'The shipping method should not be set in done deliveries.')
```

File: addons/delivery/tests/test_delivery_stock_move.py (hunk 5 above)
Comment (jva-odoo) [reply addons/delivery/tests/test_delivery_stock_move.py:192]: Why do you add in Self?
You need this on this function only!

File: addons/delivery/models/sale_order.py [hunk 7]
```diff
@@ -51,9 +51,13 @@ def _remove_delivery_line(self):
...
+        # (we have no integration of returns for now)
         self._remove_delivery_line()
         for order in self:
             order.carrier_id = carrier.id
+            if order.state in ('sale', 'done'):
+                order.picking_ids.filtered(lambda p: p.state not in ('done', 'cancel') and not p.move_ids.origin_returned_move_id).carrier_id = carrier.id
```
Comment (jva-odoo) [inline addons/delivery/models/sale_order.py:60]:
```suggestion
                pending_deliverys =  order.picking_ids.filtered(
                    lambda p: p.state not in ('done', 'cancel') and not p.move_ids.origin_returned_move_id)
                pending_deliverys.carrier_id = carrier
```

**Review by jva-odoo — CHANGES_REQUESTED** [review]
Fix it

File: addons/delivery/tests/test_delivery_stock_move.py [hunk 8]
```diff
@@ -184,3 +184,61 @@ def test_03_invoiced_status(self):
...
+        }))
+        choose_delivery_carrier = delivery_wizard.save()
+        choose_delivery_carrier.button_confirm()
+
+        # Check the carrier in picking after conform sale order
+        delivery_for_product_11 = sale_prepaid.picking_ids.filtered(lambda p: self.product_11 in p.move_ids.product_id)
+        self.assertEqual(delivery_for_product_11.carrier_id, self.normal_delivery, 'The shipping method should be set in pending deliveries.')
+
+        done_delivery = sale_prepaid.picking_ids.filtered(lambda p: p.state == "done")
+        self.assertEqual(done_delivery.carrier_id.id, False, 'The shipping method should not be set in done deliveries.')
```
Comment (jva-odoo) [inline addons/delivery/tests/test_delivery_stock_move.py:243]:
```suggestion
        self.assertFalse(done_delivery.carrier_id.id, 'The shipping method should not be set in done deliveries.')
```

File: addons/delivery/tests/test_delivery_stock_move.py [hunk 9]
```diff
@@ -184,3 +184,61 @@ def test_03_invoiced_status(self):
...
+        self.assertEqual(return_picking.carrier_id.id, False, 'The shipping method should not set in return picking')
```
Comment (jva-odoo) [inline addons/delivery/tests/test_delivery_stock_move.py:244]:
```suggestion
        self.assertFalse(return_picking.carrier_id.id, 'The shipping method should not set in return picking')
```

**Review by jva-odoo — CHANGES_REQUESTED** [review]
Just do this small fix

**Review by jva-odoo — APPROVED** [review]

**Review by ajf-odoo — COMMENTED** [review]
Perhaps the task has evolved but the branch name is not adequate.
Make also the commit message clearer.

File: addons/delivery/models/sale_order.py [hunk 10]
```diff
@@ -51,9 +51,15 @@ def _remove_delivery_line(self):
...
     def set_delivery_line(self, carrier, amount):
+        # Prevent set of the carrier when generating return picking
+        # (we have no integration of returns for now)
         self._remove_delivery_line()
         for order in self:
             order.carrier_id = carrier.id
+            if order.state in ('sale', 'done'):
+                pending_deliverys = order.picking_ids.filtered(
+                    lambda p: p.state not in ('done', 'cancel') and not p.move_ids.origin_returned_move_id)
+                pending_deliverys.carrier_id = carrier.id
```
[the remark this answers is not in the document]
Comment (kir-odoo) [reply addons/delivery/models/sale_order.py:62]: I didn't find any reference or identification to track reshipping. Would you happen to have any idea to track it?

[the remark this answers is not in the document]
Comment (ajf-odoo) [reply addons/delivery/models/sale_order.py:62]: what about location usage, picking code or incoming/outgoing moves ?

File: addons/delivery/models/sale_order.py [hunk 11]
```diff
@@ -59,6 +59,10 @@ def set_delivery_line(self, carrier, amount):
         self._remove_delivery_line()
         for order in self:
             order.carrier_id = carrier.id
+            if order.state in ('sale', 'done'):
+                pending_deliverys = order.picking_ids.filtered(
+                    lambda p: p.state not in ('done', 'cancel') and p.location_id == p.picking_type_id.warehouse_id.lot_stock_id)
+                pending_deliverys.carrier_id = carrier.id
```
Comment (Feyensv) [inline addons/delivery/models/sale_order.py:65]: The stock logic of delivery was moved to the module `stock_delivery` recently so this code now belongs in `stock_delivery` (with a new hook method in delivery if really needed) ;).

Comment (kir-odoo) [reply addons/delivery/models/sale_order.py]: Done

**Review by clesgow — COMMENTED** [review]
As this issue was first seen in august, I think it would be wiser to re-target it to `16.0` instead of current `master`.
The move from the `delivery` module to the `stock_delivery` one should be then done in the right forward-port.

**Comment by kir-odoo** [comment]
Hello, @clesgow this pr moved to https://github.com/odoo/odoo/pull/116533 here.

**Pull request closed without merging**